core/src/main/java/kafka/server/SharePartitionManager.java

```diff
@@ -348,11 +348,13 @@ public void acknowledgeShareSessionCacheUpdate(String groupId, Uuid memberId, in
             return;
         }
         ShareSession shareSession = cache.get(new ShareSessionKey(groupId, memberId));
-        // Update the session's position in the cache for both piggybacked (to guard against the entry disappearing
-        // from the cache between the ack and the fetch) and standalone acknowledgments.
-        cache.touch(shareSession, time.milliseconds());
-        // If acknowledgement is piggybacked on fetch, newContext function takes care of updating the share session epoch
-        // and share session cache. However, if the acknowledgement is standalone, the updates are handled in the if block
+        synchronized (cache) {
+            // Update the session's position in the cache for both piggybacked (to guard against the entry disappearing
+            // from the cache between the ack and the fetch) and standalone acknowledgments.
+            cache.touch(shareSession, time.milliseconds());
+            // If acknowledgement is piggybacked on fetch, newContext function takes care of updating the share session epoch
+            // and share session cache. However, if the acknowledgement is standalone, the updates are handled in the if block
+        }
```

Comment:
Looking at the `touch` method, it seems to be already synchronized on `cache` and `session`. I am not sure if we needed to be synchronized that much in code. Let's discuss tomorrow what we want to achieve and see potential thready safety areas.